Fixes & optimizations

mudcube opened the pull request:
This pull request:
- Greatly reduces garbage during exports, improving export time. 
- Fixes error thrown with font_names with underscores
- Fixes error when HTMLCanvasElement pattern is sent to C2S instance
- Adds `fill-rule` support as reported in issue #62
- Merges in `opacity` support for issue #53

mudcube committed:
Fix issue with font_names with underscores throwing an error

mudcube committed:
Merge in fix #53

via @kinimesi

mudcube committed:
Merge branch 'master' of https://github.com/gliffy/canvas2svg

mudcube committed:
Reduce garbage collections

- Remove inefficient format() for template strings
- Remove inefficient lineTo indexOf on large strings

mudcube committed:
Add `fill-rule` support

mudcube committed:
Handle multiple strokes with single path

Example:
```
	ctx.rect(75, 75, 100, 100)

	ctx.lineWidth = 30
	ctx.strokeStyle = '#0074D9'
	ctx.stroke()

	ctx.fillStyle = '#FF4136'
	ctx.fill()

	ctx.lineWidth = 10
	ctx.strokeStyle = '#FFDC00'
	ctx.stroke()
```

mudcube committed:
Add identifier to `ctx` to differentiate between native context

mudcube committed:
Fix potential infinite loop during Pattern application

The while loop was crashing my browser, so converted into a for loop. The issue is when appendChild does not remove the child from __defs.childNodes, I’m not entirely sure why this was happening, but this method also is slightly more optimized as you don’t recalculate the length upon each interation.

mudcube committed:
`patternUnits` set to `userSpaceOnUse`

When patternUnits uses the default `objectBoundingBox` patterns are parsed incorrectly for me (they do not repeat, and are incorrectly scaled). Adding this line fixes this issue.

mudcube committed:
Fix formatting for Travis

mudcube committed:
Convert Tabs to Spaces to match project formatting

mudcube committed:
Multiple fixes and improvements

- Support multiple fills w/ single path. Ex.
```js
ctx.rect(0, 0, 100, 100);
ctx.fill();
ctx.translate(50, 50);
ctx.fill();
```

- Support `gradientTransform` and `patternTransform`. Ex.
```js
ctx.rect(0, 0, 100, 100);
ctx.translate(50, 50);
ctx.fillStyle = ctx.createPattern($image, ‘repeat’)
ctx.fill();
```

- Support switching between transforms and path operations. Ex.
```js
ctx.rect(0, 0, 100, 100);
ctx.translate(50, 50);
ctx.rect(0, 0, 100, 100);
ctx.fill();
```

- Fix clearRect() so <svg> is emptied if requested cleared area covers the entire canvas area. This update takes into account the current matrix, as well as accounts for rectangles outside of canvas bounds.

mudcube committed:
Fix Save & Restore transformation test

```
ctx.translate(10, 10)
ctx.fillRect(0, 0, 10, 10)
ctx.save()
ctx.restore()
ctx.translate(10, 10)
ctx.fillRect(0, 0, 10, 10)
```

mudcube committed:
Create <def> for Paths with multiple fills or strokes

This patch creates smaller exported svgs by creating <def> entries for paths references more than one. This kicks into gear when a fill or stroke is used multiple times on a single path.

mudcube committed:
Fix Bug attempting to apply “d” attribute to “use” element

mudcube committed:
Create ID from `path` hash to cut down on duplicate <defs>

jods4 commented:
I tried this patch because the fill=evenodd bug hit me hard.

I can confirm that the evenodd fix works.

Unfortunately I also have to report that some of the other optimisations broke other stuff, I get funny results that were previously ok.

mudcube commented:
Hey jods, what new issues are you noticing with this pull request? Perhaps
I can sort them out if you give me some detailed instructions to reproduce.

On Sun, Jun 3, 2018, 11:44 AM jods <notifications@github.com> wrote:

> I tried this patch because the fill=evenodd bug hit me hard.
>
> I can confirm that the evenodd fix works.
>
> Unfortunately I also have to report that some of the other optimisations
> broke other stuff, I get funny results that were previously ok.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gliffy/canvas2svg/pull/66#issuecomment-394182503>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGMvOnmhGKR0frkC3vQfK0acdwfu-Rxks5t5C6KgaJpZM4SjoX_>
> .
>

jods4 commented:
@mudcube in my document many lines end up at the origin after your patch, they were drawn correctly before.

Unfortunately it's real documents with around 10K calls to context, that I cannot easily share or reduce.

That said, some tests are currently falling, could that be related?
I can gladly retest my docs when they are fixed.

mudcube committed:
Fix issue with ctx.transform being overwritten with DOMMatrix

mudcube committed:
Add package-lock.json

mudcube committed:
Fix for clip and style application

- Clip would sometimes report that group was not the parent of currentElement. So, instead calling `.remove()` on the element.
- Create path if does not exist when calling `getOrCreateElementToApplyStyleTo`
- Revert support switching between transforms and path operations… this needs to be rethought as it was causing other transformation issues.

mudcube committed:
Remember last test run, and rerun upon reload

mudcube committed:
Playground now updates textarea with code examples when chosen

mudcube committed:
Create <path> element only when needed

mudcube committed:
Playground now has ability to reload current test

mudcube committed:
Fixed with all files in test/example

mudcube commented:
@jods4 The transformation issue should be sorted out, let me know if you're still running into the issue. There's still a couple failing tests, but part of that is due to slight formatting changes (ex. spacing), and doesn't actually affect what's visually exported.

mudcube committed:
Prune unneeded save/restore groups

mudcube committed:
Apply transforms directly to paths/gradients

This allows the intention behind commit c228ba48dd5f58624f1dadc3381d6328a339117d to work once again. Note, this commit contains MPL code, so if this pull request is accepted an MPL license would need to be added to the repository.

jods4 commented:
@mudcube I tried again and the "all lines go to origin" problem is fixed on your master.

But it's not completely fixed.
Paths are now at their correct position, but they are way too wide. There is still a regression inside that PR.

mudcube committed:
Ensure `element` exists when setting fill-rule

mudcube committed:
update @babel/core -> 7.8.3

mudcube committed:
Convert Tabs to Spaces to match project formatting
